[IV-23-24] Objetivo 7

manuelvico0102 opened the pull request:
* [x] ¿Se han registrado todos los eventos relevantes en la aplicación, es decir, se
  han añadido comandos para registrarlos en todos ellos?
* [x] ¿Se han seguido buenas prácticas en el uso de configuración remota?
* [x] ¿Se han explicado correctamente las decisiones técnicas que se han tomado
  sobre las bibliotecas usadas?
* [x] ¿Te has asegurado de que el fichero `.env` no se pueda añadir
accidentalente al repositorio?
* [x] ¿Los logs y la configuración están en un objeto que se *inyecta* en las
clases manejadoras para su uso?
* [x] ¿*Realmente* se usan los logs en las diferentes clases en eventos
importantes?
* [x] ¿Se testea el código creado para logs y configuración, igual que
    cualquier otro código?
* [x] ¿La clase o módulo de configuración lee *siempre* las variables de entorno?

manuelvico0102 committed:
Elegido logger closes #41

Se ha elegifo pino como logger

manuelvico0102 committed:
Elegido sistema de configuración closes #42

Se ha elegido dotenv como sistema de configuracion

manuelvico0102 committed:
Se ha realizado la configuración closes #44

Se ha configurado para que coja las variables del archivo .env, si no toma un valor por defecto las variables

manuelvico0102 committed:
Se ha configurado el logger closes #43

Se ha implementado el logger con la configuración

manuelvico0102 committed:
Se ha añadido la llamadas al logger closes #45

Las clases tiene la llamada al logger para registrar diferentes estados

manuelvico0102 committed:
Implementado test para log closes #46

Se ha creado un test para comprobar el funcionamiento del logger, desde la creacion de un candidato

manuelvico0102 committed:
Actualizados ficheros de dependencias #47

Se ha actualizado el fichero de dependencias para el logger y la configuracion

manuelvico0102 committed:
Actualizado Readme closes #48

Se ha actualizado el readme con las referencias para el logger y el sistema de configuración

manuelvico0102 committed:
Actualizado iv.yaml #49

Se ha actualizado el iv.yaml para indicar el fichero de configuracion

manuelvico0102 committed:
Actualizado .gitignore closes #49

Se ha actualizado el .gitignore para que no se suba el .env

manuelvico0102 committed:
Actualizado Readme closes #48

Se ha actualizado el readme con las referencias para el logger y el sistema de configuración

manuelvico0102 committed:
Modificada clase configuracion closes #44

Se ha cambiado la clase configuración para tener un get generico

manuelvico0102 committed:
Añadido logger por pantalla closes #43

Se ha añadido otro stream para mostrar por pantalla el logger y que no sea solo guardando en el archivo

manuelvico0102 committed:
Añadido variable para github #44

se ha indicado la variable de github para que las lea ya que no esta el archivo .env

manuelvico0102 committed:
Añadido log_file_path a partir de .env #43

Se ha añadido el codigo para tomar el valor de log_file_path de .env

manuelvico0102 committed:
Añadido LOG_LEVEL y LOG_FILE_PATH #47

Se ha añadido lo necesario para pasar los test ci

manuelvico0102 committed:
Eliminado caracteres innecesarios #49

Habian unos carácteres al final que han hecho que no pasasen los test

pluque01 commented on src/candidato.ts:
```diff
@@ -9,20 +10,23 @@ export class Candidato {
     private idiomas: Idiomas;
     private lenguajesDeProgramacion: LenguajesDeProgramacion;
+    
+    private logger : any = loggerConfigurado.logger;
```
No le encuentro mucho sentido a que incluyas una variable de instancia para el logger. El logger debería ser un módulo externo que se pueda importar donde definas los métodos de la clase.

manuelvico0102 replied:
Solucionado en dad91a4ecc99ec8114ba4aca622b08384a79442e

pluque01 commented on src/oferta.ts:
```diff
@@ -1,12 +1,15 @@
 class Oferta {
     private id: string;
     private candidatos: Candidato[];
+    private logger : any = loggerConfigurado.logger;
```
Aquí lo mismo que te he comentado antes

manuelvico0102 replied:
Solucionado en dad91a4ecc99ec8114ba4aca622b08384a79442e

pluque01 reviewed (commented):
Hola!
Yo lo veo bien, pero cambiaría lo que te he comentado.
Un saludo!

manuelvico0102 committed:
Eliminado variable de instancia closes #45

Se han eliminado las variables de instancias de las clases para los logger

pluque01 reviewed (approved):
Lo veo bien! 👍

manuelvico0102 commented:
@JJ Listo para revisión

JJ commented on docs/eleccion_logger.md:
```diff
@@ -0,0 +1,25 @@
+
+## Bibliotecas de logs
+
+Existen diferentes bibliotecas de logs que podemos utilizar con Node.js, entre los que se encuentran: Winston, Pino y Log4js.
```
Me temo que a estas alturas convendría que buscárais un poco más bibliotecas de logging. Mirando en npmjs es fácil encontrarlas, incluso específicas de typescript.

manuelvico0102 replied:
Solucionado en 2a3f239a268ff718d061eba471c2f830702cacc3

JJ commented on docs/sistema_configuracion.md:
```diff
@@ -0,0 +1,24 @@
+
+## Elección del sistema de configuración
+
+Finalmente, para mi proyecto he escogido dotenv como sistema de configuración. Esto se debe a que dotenv es una biblioteca de configuración que permite cargar variables de entorno desde un fichero .env. Además, dotenv tiene una puntuación alta en Snyk Advisor y su mantenimiento es activo. Sin embargo, podría haber escogido nconf ya que tiene una puntuación más alta en Snyk Advisor, pero he escogi ...
```
¿Has considerado alguna de las otras que permitan configurar también usando ficheros JSON u otros dispositivos?

manuelvico0102 replied:
Solucionado en 49628ee05f1a85879e61f633f99fc4b12f515f15

JJ commented on docs/sistema_configuracion.md:
```diff
@@ -0,0 +1,24 @@
+
+- **Mantenimiento**: La biblioteca de configuración debe estar mantenida, para ello también se tendrá en cuenta lo que aparezca en Snyk Advisor.
+
+- **Compatibilidad**: La biblioteca de configuración debe ser compatible para nuestro proyecto.
```
¿Cómo vas a considerar alguna que no sea compatible?
Aparte, ¿no tendrías que tener en cuenta alguna cosa más específica de tu proyecto? Por ejemplo, [este fichero](https://github.com/manuelvico0102/easySelect/blob/main/src/types.ts), ¿no sería más sencillo que estuviera en un fichero de configuración para ser fácilmente modificable en vez de en Typescript? (al menos `NivelIdioma` y `NivelLenguajeProgramacion`)

JJ commented on docs/sistema_configuracion.md:
```diff
@@ -0,0 +1,24 @@
+
+## Opciones
+
+Existen diferentes bibliotecas de configuración que podemos utilizar, entre los que se encuentran: 
```
Por otro lado, igual que antes, ¿has buscado en npmjs módulos específicos de TypeScript para configuración? ¿O incluso otros módulos aunque sean solo de javascript? Hay bastantes que no están en la lista de abajo

manuelvico0102 replied:
Solucionado en 49628ee05f1a85879e61f633f99fc4b12f515f15

JJ commented on src/utils/configuracion.ts:
```diff
@@ -0,0 +1,20 @@
+        const value = process.env[clave];
+
+        if (!value) {
+            const githubEnvClave = process.env[`ENV_${clave}`];
```
Las claves de github no usan ese prefijo. Y por otro lado, también sería conveniente que consideraras otras bibliotecas de configuración.

manuelvico0102 replied:
Solucionado en c316ccd56cf03a201b81a4751fc1f627cf101130

JJ reviewed (changes requested):
Como sabes, en las entregas posteriores se pide un poco más que en las primeras; no es difícil que encuentres otras bibliotecas diferentes, usando criterios de elección un poco más específicos.

manuelvico0102 committed:
Logger de typescript closes #41

Se han indicado algunos log de typescript y se han añadido algo más de información, los logger de typescript se han mencionado pero no se han elegido por su baja puntuación en SnykAdvisor

manuelvico0102 committed:
Añadido sistema de configuración typescript closes #42

Se ha añadido cosmiconfig, sin embargo se ha mantenido dotenv

manuelvico0102 committed:
Cambiado prefijo para clave github closes #44

JJ commented:
¿Qué queda por aquí?

manuelvico0102 commented:
Me quedó resolver la sugerencia: "Aparte, ¿no tendrías que tener en cuenta alguna cosa más específica de tu proyecto? Por ejemplo, [este fichero](https://github.com/manuelvico0102/easySelect/blob/main/src/types.ts), ¿no sería más sencillo que estuviera en un fichero de configuración para ser fácilmente modificable en vez de en Typescript? (al menos NivelIdioma y NivelLenguajeProgramacion)", pero no he tenido mucho tiempo para ponerme a resolverlo

JJ commented:
Inténtalo al menos, ¿no?
